### PR TITLE
feat (graphql): Datahub GMS Graphql Api Application for Querying Dataset

### PR DESCRIPTION
**Pull request by arunvasudevan**
Datahub GMS GraphQL API service adds a /graphql endpoint to the gms service.

Here is the highlevel overview of the Architecture

Datahub GMS GraphQL API

To verify the changes in a docker use the below command
`docker-compose -p datahub -f docker-compose.yml up datahub-gms-graphql-api`

Endpoint: http://localhost:8091/graphql

`GraphQL Query`:

```
{
  dataset(urn: "urn:li:dataset:(urn:li:dataPlatform:foo,bar,PROD)") {
    urn
    platform
    name
    origin
    description
    platformNativeType
    uri
    tags
    ownership {
      owners {
        owner {
          username
          urn
          info {
            displayName
            email
            fullName
            manager {
              urn
            }
          }
          editableInfo {
            aboutMe
            skills
          }
        }
        type
        source {
          url
        }
      }
      lastModified {
        actor
      }
    }
    created {
      actor
    }
    lastModified {
      actor
    }
  }
}
```

## Checklist
- [ X] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ X] Tests for the changes have been added/updated (if applicable)
- [ X] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -56,6 +56,8 @@ project.ext.externalDependency = [
     'gmaTestModelsDataTemplate': "com.linkedin.datahub-gma:test-models-data-template:$gmaVersion",
     'gmaValidators': "com.linkedin.datahub-gma:validators:$gmaVersion",
     'graphqlJava': 'com.graphql-java:graphql-java:16.1',
+    'graphqlJavaSpringWebMvc': 'com.graphql-java:graphql-java-spring-boot-starter-webmvc:2.0',
+    'graphiqlSpring': 'com.graphql-java:graphiql-spring-boot-starter:5.0.2',
```
Comment (jjoyce0510) [inline build.gradle:60]: This is used for graph visualization, correct? Is this necessary to include in DH?

File: datahub-gms-graphql-api/build.gradle [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    compile project(':datahub-graphql-core')
+    compile (externalDependency.springBootStarterWeb) {
+        exclude module: "spring-boot-starter-tomcat"
+    }
+    compile externalDependency.graphqlJava
+    compile externalDependency.antlr4Runtime
+    compile externalDependency.antlr4
+    compileOnly externalDependency.lombok
+    annotationProcessor externalDependency.lombok
+    compile externalDependency.graphqlJavaSpringWebMvc
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/build.gradle:16]: minor organization nitpick: Place these compile statements above the `annotationProcessor` and `compileOnly` statements

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/DatahubGmsGraphQLApiApplication.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package com.linkedin.datahub.graphql.api;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.data.elasticsearch.ElasticsearchAutoConfiguration;
+import org.springframework.boot.autoconfigure.elasticsearch.rest.RestClientAutoConfiguration;
+
+@SuppressWarnings("checkstyle:HideUtilityClassConstructor")
+@SpringBootApplication(exclude = {ElasticsearchAutoConfiguration.class, RestClientAutoConfiguration.class})
+public class DatahubGmsGraphQLApiApplication {
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/DatahubGmsGraphQLApiApplication.java:10]: Nit: Are 'Daahub' & 'Api' necessary as part of the name here?

Consider `GmsGraphQLApplication` or something similar

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/GmsGraphQLProvider.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    private GraphQL graphQL;
+    private GraphQLEngine graphQLEngine;
+
+    @Bean
+    public GraphQL graphQL() {
+        return graphQL;
+    }
+
+    @Bean
+    public GraphQLEngine graphQLEngine() {
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/GmsGraphQLProvider.java:25]: Do we need to explicitly provide the `GraphQLEngine` in addition to the `GraphQL` instance, or is the`GraphQL` instance enough?

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        DataLoaderRegistry register = graphQLEngine.createDataLoaderRegistry(graphQLEngine.get_dataLoaderSuppliers());
+
+        ExecutionInput executionInput = ExecutionInput.newExecutionInput()
+            .query(invocationData.getQuery())
+            .operationName(invocationData.getOperationName())
+            .variables(invocationData.getVariables())
+            .dataLoaderRegistry(register)
+            .context(queryContext)
+            .build();
+        return graphQL.executeAsync(executionInput);
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java:45]: Are we unable to leverage the `execute` API that exists within `GraphQLEngine`? 

It seems that `GraphQLEngine` is only being used here to help initialize the DataLoaders, is that the correct interpretation?

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 6]
```diff
@@ -31,6 +33,7 @@
  *
  * <p>In addition, it provides a simplified 'execute' API that accepts a 1) query string and 2) set of variables.
  */
+@Data
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:36]: Please don't make this a @Data class. I'd prefer everything that we expose from here to be explicit. 

Instead, we can just add `getter` methods for the information you need to retrieve

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 7]
```diff
@@ -152,7 +155,7 @@ public GraphQLEngine build() {
         }
     }
 
-    private DataLoaderRegistry createDataLoaderRegistry(final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers) {
+    public DataLoaderRegistry createDataLoaderRegistry(final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers) {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:158]: Would it make sense to expose a method called `getDataLoaderRegistry` that returns a cached version of a registry that was already created? 

That way, the caller does not need to explicitly pass in the `dataLoaderSuppliers`

File: datahub-graphql-core/src/main/resources/gms.graphql [hunk 8]
```diff
@@ -3,7 +3,7 @@ scalar Long
 
 schema {
     query: Query
-    mutation: Mutation
+#    mutation: Mutation
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/resources/gms.graphql:6]: Is there a reason we need to comment this out? It will break the `datahub-frontend` server (which extends the Mutation type) if we do

File: settings.gradle [hunk 9]
```diff
@@ -25,4 +25,6 @@ include 'metadata-models-ext'
 include 'metadata-models-generator'
 include 'metadata-testing:metadata-models-test-utils'
 include 'metadata-testing:metadata-test-utils'
-include 'metadata-utils'
\ No newline at end of file
+include 'metadata-utils'
+include 'datahub-gms-graphql-api'
```
Comment (jjoyce0510) [inline settings.gradle:29]: Can we consider alternative names for this module? When I read 'api' in a Java project I tend to think of a module containing public interfaces / the publicly accessible portions of a library.

Basically we should try to capture the idea that the module contains a deployable service using the name. 

Alternatives could be 

- datahub-gms-graphql-spring-boot
- datahub-gms-graphql-service
- datahub-gms-graphql-server

Or something similar. Interested to hear your thoughts!

**Review by jjoyce0510 — CHANGES_REQUESTED** [review]
Overall this is looking really good! A few comments.

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/DatahubGmsGraphQLApiApplication.java (hunk 3 above)
Comment (arunvasudevan) [reply datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/DatahubGmsGraphQLApiApplication.java:10]: As always naming is the hard part, thanks for providing suggestions here...

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java (hunk 6 above)
Comment (arunvasudevan) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:36]: Makes sense..Instead I can just `@Getter`. Intention is just to reduce the noise here.

**Comment by mars-lan** [comment]
@jjoyce0510 Is the goal to migrate the new React FE to using this instead of the Play mid-tier graphql endpoint? Maintaining 2 stacks would be difficult.

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java (hunk 5 above)
Comment (arunvasudevan) [reply datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java:45]: As shown in this example https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/#schema `graphql-java-spring-boot-starter-webmvc` dependency works with the `GraphQL` bean.

Main reason to use the `GraphQLEngine` was to override the default `GraphQLContext` with `QueryContext` thereby supplying the dataProviders.

If we have to use the `GraphQLEngine` execute In my opinion i have to implement the low level details which is taken care by `graphql-java-spring-boot-starter-webmvc` (https://github.com/graphql-java/graphql-java-spring/blob/master/graphql-java-spring-webmvc/src/main/java/graphql/spring/web/servlet/components/GraphQLController.java)

File: settings.gradle (hunk 9 above)
Comment (arunvasudevan) [reply settings.gradle:29]: `datahub-gms-graphql-service` makes sense to me.

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/GmsGraphQLProvider.java (hunk 4 above)
Comment (arunvasudevan) [reply datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/GmsGraphQLProvider.java:25]: As`GraphQLEngine` is used by the class `QueryGraphQLInvocation` it needs to be provided to be Autowired in the Invocation class. 
Let me know if there is a better way to do this.

File: datahub-graphql-core/src/main/resources/gms.graphql (hunk 8 above)
Comment (arunvasudevan) [reply datahub-graphql-core/src/main/resources/gms.graphql:6]: Currently the mutation type has no fields I get a compilation error.

```
nested exception is graphql.schema.validation.InvalidSchemaException: invalid schema:
"Mutation" must define one or more fields.
```

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java (hunk 7 above)
Comment (arunvasudevan) [reply datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:158]: I agree! 👍

**Comment by arunvasudevan** [comment]
@jjoyce0510 Resolved most of the comments let me know what you think.

File: settings.gradle (hunk 9 above)
Comment (arunvasudevan) [reply settings.gradle]: On second thought considering to rename it as just `datahub-gms-graphql`. Any thoughts?

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Autowired
+    GraphQLEngine graphQLEngine;
+
+    public static final String APPNAME = "DatahubGmsGraphQLApp";
+
+    @Override
+    public CompletableFuture<ExecutionResult> invoke(GraphQLInvocationData invocationData, WebRequest webRequest) {
+        QueryContext queryContext = new SpringQueryContext(true, APPNAME);
+        DataLoaderRegistry register = graphQLEngine.createDataLoaderRegistry(graphQLEngine.get_dataLoaderSuppliers());
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java]: nit: registry

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java (hunk 5 above)
Comment (jjoyce0510) [reply datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java]: What I meant to suggest is... Why not do something like the following instead:

```
return graphQLEngine.execute(invocationData.getQuery(), invocationData.getVariables(), queryContext); 
```

If you did this, you likely wouldn't need _both_ `GraphQL` and `GraphQLEngine` in this class, just the latter.

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    GraphQL graphQL;
+
+    @Autowired
+    GraphQLEngine graphQLEngine;
+
+    public static final String APPNAME = "DatahubGmsGraphQLApp";
+
+    @Override
+    public CompletableFuture<ExecutionResult> invoke(GraphQLInvocationData invocationData, WebRequest webRequest) {
+        QueryContext queryContext = new SpringQueryContext(true, APPNAME);
```
Comment (jjoyce0510) [inline datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java]: You are passing in APPNAME here... Soon, we'll have to understand how to authorize specific actions (say, mutating the ownership of a Dataset) against a concrete human actor (like a corpUser) at some point. To do so, we'll want to have the context of the _actual_ actor (eg. which user is making the change) to perform authorization against. 

Perhaps we can come up with some standardized "super" actor that we will always let through, and if you really cannot pass the actual actor context you can use that for your use case

This may mean that in the future you may need access to that context.

File: datahub-graphql-core/src/main/resources/gms.graphql (hunk 8 above)
Comment (jjoyce0510) [reply datahub-graphql-core/src/main/resources/gms.graphql:6]: Ah got it, okay

File: settings.gradle (hunk 9 above)
Comment (jjoyce0510) [reply settings.gradle]: I'm concerned people will wonder what the difference is between `datahub-gms-graphql` and `datahub-gms-graphql-core` if there is no additional qualifier. 

I think it's important to include something in the name suggesting "this is just a graphql-compatible deployable"

**Comment by jjoyce0510** [comment]
> Is the goal to migrate the new React FE to using this instead of the Play mid-tier graphql endpoint? Maintaining 2 stacks would be difficult.

@mars-lan That's not currently part of the goals. Both our use cases (Play server) and Arun's will share the same base graph schema, business logic required for resolving entities from GMS, but the final deployable layer will remain separate for now. The fact that we are sharing the meat of the code should reduce the cost of maintaining separate deployment containers.

I think we can continue to evaluate whether it makes sense to consolidate completely to what Arun's working on, but we'd need to port some additional business logic from the existing `datahub-frontend` pertaining to authentication (the login module, auth cookie creation, etc), which @arunvasudevan has made clear he does not need for his use case. 

TLDR; Because we are sharing the graphql core code maintenance cost should be small. It's not out of the question in the future, but it's not in the current plan.

File: datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java (hunk 5 above)
Comment (arunvasudevan) [reply datahub-gms-graphql-api/src/main/java/com/linkedin/datahub/graphql/api/QueryGraphQLInvocation.java]: Got it!

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 12]
```diff
@@ -33,18 +34,20 @@
  *
  * <p>In addition, it provides a simplified 'execute' API that accepts a 1) query string and 2) set of variables.
  */
-@Data
+@Getter
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:37]: Can we just implement the getter methods manually, instead of using lombok? In the future if we added a new `private final` member variable we'd automatically generate the getter, which may not be what we want (eg. if we want to hide internal impl details).

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 13]
```diff
@@ -33,18 +34,20 @@
...
 
     private GraphQLEngine(@Nonnull final List<String> schemas,
                           @Nonnull final RuntimeWiring runtimeWiring,
-                          @Nonnull final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers) {
+                          @Nonnull final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers,
+                            @NonNull final DataLoaderRegistry registry) {
 
         _dataLoaderSuppliers = dataLoaderSuppliers;
-
+        _registry = registry;
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:50]: Why are we adding this new private variable?

The registry should be re-created on each request. From the `graphql-java` official documentation https://www.graphql-java.com/documentation/v15/batching/ ( see "Per Request Data Loaders" section ) 

>         //
        // DataLoaderRegistry is a place to register all data loaders in that needs to be dispatched together
        // in this case there is 1 but you can have many.
        //
        // Also note that the data loaders are created per execution request
        //
        DataLoaderRegistry registry = new DataLoaderRegistry();
        registry.register("character", characterDataLoader);

        ExecutionInput executionInput = newExecutionInput()
                .query(getQuery())
                .dataLoaderRegistry(registry)
                .build();



For that reason, I don't include it on creation of the `GraphQLEngine`, but instead recreate for each request using the `dataLoaderSuppliers`.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 14]
```diff
@@ -67,18 +70,13 @@ private GraphQLEngine(@Nonnull final List<String> schemas,
     public ExecutionResult execute(@Nonnull final String query,
                                    @Nullable final Map<String, Object> variables,
                                    @Nonnull final QueryContext context) {
-        /*
-         * Init DataLoaderRegistry - should be created for each request.
-         */
-        DataLoaderRegistry register = createDataLoaderRegistry(_dataLoaderSuppliers);
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java:70]: This should be left as is, with the registery* (typo) created on each execution

File: datahub-gms-graphql-service/src/main/java/com/linkedin/datahub/graphql/service/GmsGraphQLProvider.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import graphql.GraphQL;
+
+@Component
+public class GmsGraphQLProvider {
+
+    private GraphQL graphQL;
+    private GraphQLEngine graphQLEngine;
+
+    @Bean
+    public GraphQL graphQL() {
```
Comment (jjoyce0510) [inline datahub-gms-graphql-service/src/main/java/com/linkedin/datahub/graphql/service/GmsGraphQLProvider.java:20]: This can be removed now, as it is unused, right?

Comment (arunvasudevan) [reply datahub-gms-graphql-service/src/main/java/com/linkedin/datahub/graphql/service/GmsGraphQLProvider.java:20]: As shown here - https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/#schema
`graphQL` bean is needed by the `graphql-java-spring-boot-starter-webmvc` dependency.

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 16]
```diff
@@ -34,20 +32,16 @@
  *
  * <p>In addition, it provides a simplified 'execute' API that accepts a 1) query string and 2) set of variables.
  */
-@Getter
 public class GraphQLEngine {
 
     private final GraphQL _engine;
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java]: This is actually misnamed... We can change it to `graphQL`

File: datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java [hunk 17]
```diff
@@ -161,7 +149,19 @@ public Builder configureRuntimeWiring(final Consumer<RuntimeWiring.Builder> buil
...
+
+    private DataLoaderRegistry createDataLoaderRegistry(final Map<String, Supplier<DataLoader<?, ?>>> dataLoaderSuppliers) {
+        final DataLoaderRegistry registry = new DataLoaderRegistry();
+        for (String key : dataLoaderSuppliers.keySet()) {
+            registry.register(key, dataLoaderSuppliers.get(key).get());
         }
+        return registry;
+    }
+
+    public GraphQL get_engine() {
```
Comment (jjoyce0510) [inline datahub-graphql-core/src/main/java/com/linkedin/datahub/graphql/GraphQLEngine.java]: Please follow camelcase for the java function names... 

`getGraphQL`

**Comment by jjoyce0510** [comment]
lgtm! Thanks for your contribution Arun!

**Review by shirshanka — APPROVED** [review]
Thanks for addressing all the feedback @arunvasudevan! This LGTM.
